pkg/prometheus/promcfg.go

```diff
@@ -2499,11 +2499,79 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 	if len(sc.Spec.KubernetesSDConfigs) > 0 {
 		configs := make([][]yaml.MapItem, len(sc.Spec.KubernetesSDConfigs))
 		for i, config := range sc.Spec.KubernetesSDConfigs {
-			configs[i] = []yaml.MapItem{
-				{
-					Key:   "role",
-					Value: strings.ToLower(string(config.Role)),
-				},
+			assetStoreKey := fmt.Sprintf("scrapeconfig/%s/%s/kubernetessdconfig/%d", sc.GetNamespace(), sc.GetName(), i)
+			if config.APIServer != nil {
+				configs[i] = []yaml.MapItem{
+					{
+						Key:   "api_server",
+						Value: config.APIServer,
+					},
+				}
+			}
+			configs[i] = append(configs[i], yaml.MapItem{
+				Key:   "role",
+				Value: strings.ToLower(string(config.Role)),
+			})
+
+			configs[i] = cg.addBasicAuthToYaml(configs[i], assetStoreKey, store, config.BasicAuth)
+			configs[i] = cg.addSafeAuthorizationToYaml(configs[i], fmt.Sprintf("scrapeconfig/auth/%s/%s/kubernetessdconfig/%d", sc.GetNamespace(), sc.GetName(), i), store, config.Authorization)
+			configs[i] = cg.addOAuth2ToYaml(configs[i], config.Oauth2, store.OAuth2Assets, assetStoreKey)
+
+			if config.ProxyConfig != nil {
+				if config.ProxyConfig.ProxyURL != nil {
+					configs[i] = cg.WithMinimumVersion("2.43.0").AppendMapItem(configs[i], "proxy_url", *config.ProxyConfig.ProxyURL)
+				}
+
+				if config.ProxyConfig.NoProxy != nil {
+					configs[i] = cg.WithMinimumVersion("2.43.0").AppendMapItem(configs[i], "no_proxy", *config.ProxyConfig.NoProxy)
+				}
+
+				if config.ProxyConfig.ProxyFromEnvironment != nil {
+					configs[i] = cg.WithMinimumVersion("2.43.0").AppendMapItem(configs[i], "proxy_from_environment", *config.ProxyConfig.ProxyFromEnvironment)
+				}
+
+				if config.ProxyConfig.ProxyConnectHeader != nil {
+					proxyConnectHeader := make(map[string]string, len(config.ProxyConfig.ProxyConnectHeader))
+
+					for k, v := range config.ProxyConfig.ProxyConnectHeader {
+						value, err := store.GetKey(ctx, sc.GetNamespace(), monitoringv1.SecretOrConfigMap{
+							Secret: &v,
+						})
+
+						if err != nil {
+							return configs[i], fmt.Errorf("failed to read %s secret %s: %w", v.Name, jobName, err)
+						}
+
+						proxyConnectHeader[k] = value
+					}
+
+					configs[i] = cg.WithMinimumVersion("2.43.0").AppendMapItem(configs[i], "proxy_connect_header", stringMapToMapSlice(proxyConnectHeader))
+				}
+			}
+
+			if config.FollowRedirects != nil {
+				configs[i] = append(configs[i], yaml.MapItem{
+					Key:   "follow_redirects",
+					Value: config.FollowRedirects,
+				})
+			}
+
+			if config.EnableHttp2 != nil {
+				configs[i] = append(configs[i], yaml.MapItem{
+					Key:   "enable_http2",
+					Value: config.EnableHttp2,
+				})
+			}
+
+			if config.TLSConfig != nil {
+				configs[i] = addSafeTLStoYaml(configs[i], sc.GetNamespace(), *config.TLSConfig)
+			}
+
+			if config.Namespaces != nil {
+				configs[i] = append(configs[i], yaml.MapItem{
```

Comment:
I'd rather iterate over every field than assuming an identical serialization between the CRD and Prometheus configuration.